README.md

# Power Quality on a Pi


## Synopsis

 

In order to better understand the power grid and related systems, I built a small, low cost device that allows for the measurement of voltage, frequency, and current fluctuations with very high precision. The sensitivity of the device is so great that it can also be used to "listen" to events on the power grid.
With this device I can run a range of experiments and collect data to gain familiarity.

The term power quality is an industry term that refers to the quality of electricity, namely how well all aspects of the electricity are within acceptable ranges.

In this article I will introduce the experiment, show how it is built, and then I’ll share my observations and what I have learned so far.

Thanks to the folks in /r/signalidentification for showing interest and encouraging me to write this article.

##  System Overview

![System Overview](images/acpq.png "Power Quality System Overview")


## Hardware

### Voltage Probe Parts list

Hardware can vary widely. You can use your own equipment but your mileage may vary. For example if you use the sound card in a computer you may find the power supply to the sound card drifts and introduces some noise, though likely not a deal breaker.

1. Raspberry pi 2b or newer
1. Class 10 sd card (using a good fast SD card makes all the difference in a pi)  16 gig or more. Raspbian
1. 12v AC "wall wart"
1. Two 1k resistors
1. One 100 ohm resistor
1. A 3.5mm male connector
1. A Syba USB sound adapter (the black one)
1. An Alisten 6 Port USB-C and USB Charger (35 watt 5v USB charging station)


### Assembly Instructions

Follow the above schematic to build the voltage probe.

**If anybody is good with electronics please offer suggestions for improving the probe and voltage divider. I have a hunch more dynamic range in the readings can be had with a more optimized voltage divider or maybe something else altogether.**

## Software

### Requirements

The two biggest challenges with installing software are:
1. The custom gnuradio module gr-powerquality which installs the getfreqcpp block
1. The acpq.py and other python scripts have various Python module dependencies.


### Installation

Last tested against Rasbian 2017 debian 8.0

Install yaml library dependency, used to read acpq.py config files.
```
apt-get install libyaml-dev
```

Install other required Python Modules
```
apt-get install python-pip python-pysocks
sudo pip install elasticsearch certifi PyYAML
```
Install the gr-powerquality gnuradio module
```
** installing gnuradio on Raspi
# apt-get install gnuradio-dev
# apt-get install sudo apt-get install cmake libboost-all-dev libcppunit-dev swig doxygen liblog4cpp5-dev python-scipy
$ mkdir git; cd git
$ git clone https://github.com/regulatre/gr-powerquality
$ cd gr-powerquality
$ mkdir build ; cd build
$ cmake ../; make; sudo make install; sudo ldconfig
```

Now that you've installed the custom gnuradio block we can use it from python applications such as acpq.py.

(If anybody wants to help me create a .deb it might save everybody the time of compiling)

### What's in the software repo

1. The repo itself is the custom gnuradio "out of tree" module called gr-powerquality, which contains the getfreqcpp block.
    1. The gnuradio block requires a few steps to get installed (install gnuradio from source to get all the required dependencies, then in the module directory create a build directory, cd to it, cmake ../; make; sudo make install).
1. acpq.py
    1. Script location: exanmples/acpq.py
    1. Averages voltage using alpha parameter set in config file. Readings every 1 second to elasticsearch are sufficient. Min/max spikes are also detected and reported with each reading.
    1. Uses the getfreqcpp custom block to measure precise frequency.
    1. Can/should be run as a service.
    1. Provides raw "tap" (TCP socket) to support on-demand testing such as custom gnuradio flow graphs. The streamer connects to this port to generate its audio stream.
    1. Suitable to run on a raspberry pi. I have one raspberry pi 3 with four probes, each with an associated acpq.py service running (use -c parameter and four config files).
1. Streamer
    1. Script name: examples/stream_acSoundOnly-stereo (bash script)
    1. Bash script that uses a simple flow graph to mute the fundamental and print the output as wav and then that output is piped to vlc for streaming.
    1. Listen live!
    1. Not suitable to run a pi 2 or 3 because the transcoding process (converting wav to mp3) takes a lot of CPU power - more than the Pi can handle. It might run but lots of samples will be lost, making it useless. Run it on a PC.
1. Recorder
    1. Script name: examples/circularBufferOfVlcRecordings.sh
    1. Continuously records 5 minute mp3 clips.
    1. Smart enough to ensure no lost time.
    1. Purges files older than n days.
    1. Automatic spectrogram if sox compiled with mp3 support. See SoX section below for sample spectrogram


#### calibration

Each probe will vary slightly from the next. Calibrating the audio gain of the sound driver allows us to amplify or attenuate the incoming signal to get it into useful bounds and avoid data loss due to clipping or over driving the sound card.

Calibration can be disrupted by any of the following
1. Voltage changes to the USB sound card. Use a very stable power source.
1. Other applications messing with sound card gain
1. Adjustments to the voltage probe circuit.
1. Potentially even moving the wires of the voltage probe - coiling it versus straight; wires near other power wires. etc.

Calibration Procedure:

1. Monitor the audio with gnuradio scope if possible otherwise listen. (TODO: Someone should make a python utility to do this step)
1. Turn off agc on the mic input. Built-in AGC maintains a gain level on the input that is too high for our needs and actually leads to clipping (which is bad for readings).
1. Set the gain as low as possible to maximize sensitivity to voltage spikes. Setting the gain too high results in clipping of the waveform.
    1. run arecord -l to get a list of record devices and find the device where your probe is connected
    1. run alsamixer -D hw:1 to start alsamixer controls against device #1, change device number if needed.
    1. use arrow keys to highlight Auto gain control and press M to disable ("mute") AGC.
    1. Press F4 to proceed to the capture device gain setting. 6 or 12 is a good starting point usually.
    1. hit escape when finished in alsamixer.
    1. To make changes persistent across reboots, type alsactl store 1  (change the 1 to whichever device number your probe is connected to)
1. Clipping looks obvious on a scope view because the otherwise nice smooth sine wave goes flat at the crests and troughs. Listening to it you will hear a jagged or crackling sound when its clipping. (TODO: insert image if alsamixer best settings)
1. Any changes to the curcuit or gain settings will require a recalibration of the acpq config if you want to ensure accurate voltage readings.
1. In terms of audio monitoring of the line voltage, the agc block in the flowgraph compensates nicely for gain variance so as long as the waveform isn't clipping we are in good shape.

### Voltage Measurement

To measure voltage we have to have a stable measurement environment and an accurate calibration factor

(TODO: describe the acpq config file)

To ensure a stable measurement environment, make sure your power supply is providing steady voltage, namely to the usb sound card. For a pc with usb sound card it's hard to avoid but with a raspberry pi you can better manage the power.

I have had to chase down many power issues. little cheap 5v chargers are notorious for introducing all sorts of unwanted noise and variance and messing up the measurements. Also a big surprise to me was that USB power battery packs such as those used for charging phones also introduce weird glitches to the power (when they are charging at the same time as being used. Charge circuits do weird things). Fades, dropouts, spikes, the power coming from them while on the charger is invariably quite dirty. Stick to supercaps for filtering. Run on battery power to get baseline readings since battery power is steady and free of harmonics. The last thing you want is to chase down signals that you think are coming from the grid but actually are being caused by dirty power to the soundcard.

Best results come from using this:

Raspberry pi lowered by <insert anker or other charver name a d link> as well as a small bank of supercapacitors in parallel with the voltage source, for filtering purposes.

Computer power varies wildly and the soundcard measurements from a computer are thus quite wacky based on what I’ve seen.

Calibrating the values in acpq config file is a matter of dividing actual line voltage by the raw RMS reading from the sound card (printed by acpq) which is a floating point between 0 and 1. You will need a reliable voltage reference too - a multimeter can be used or a kill-a-watt. To see the current voltage compared to the raw sensor readings.

(TODO: insert a screencast showing the process)

### Current Measurement

You can configure acpq.py to post current measurements to elasticsearch (or the screen) by tweaking the config.yml parameters to tell the script that the probe is a current probe. A current probe is a CT coil with a voltage divider similar to the one we used for voltage but the ct coil is lower voltage so the resistors are different values. You might only need a shunt resistor across the coil.

(TODO: look at ct probe wires and see what resistance I used for the shunt)

## Elasticsearch "ELK"

Elasticsearch is an open source data warehousing platform perfectly suited for big data storage and analysis. Elk refers to the trio of elasticsearch, logstash (which we dont use), and kibana, which lets you mine the data and produce wonderful visualizations and dashboards.

Kibana Dashboard

![ELK Dashboards are cool](images/elk/pq-dashboard.png)

Voltage Graph

![Elk Voltage Graph](images/elk/pq-volts.png)

## SoX Spectrograms
Spectrograms can be generated by the sox utility. Spectrograms are extremely useful for analyzing a long period of data for anomalies visually.

The recorder script automatically runs sox to generate a spectrogram for each recording.


HISTORY.md
### **How it started**

Cyber security and the power grid are intertwined; we rely on the power grid now more than ever. It touches every aspect of our lives, from charging our phones to charging our cars. Its continuous reliability and resilience is critical to our day to day lives. Not only are faults and outages a significant concern but so are other things - attacks. Of course the obvious scenario of a  power plant being "hacked" is well known but consider this: As a shared medium between many of our private devices, it (the outlets in your house or business) presents a path for devices to both communicate covertly (covert channel)  and eavesdrop on other devices and their operations (side-channel). More and more devices are connected every day. Today the risks are low and may not be of any concern to most people, but I find the topic fascinating. I want to learn more and get ahead of the risks *before* they arise so that *if* they do arise, I will be well informed and ready to help. 

### **Early Goals**

In the beginning, I had simple goals. I wanted to measure the voltage and frequency of the power from an outlet with as best accuracy as possible. I also thought it would be interesting to visualize any minute harmonics also riding on the AC waveform but didn’t know how yet. I knew I needed an AD converter (to convert analog measurements into digital form, for  computer analysis). I considered using a HackRF or RTLSDR but those are tailored for extremely wide bandwidth with lower dynamic range bits. And I know that because of the nature of the power system (power gets stepped down repeatedly at each stage of the system until it gets to the home) that most higher frequency harmonics would be eliminated due to inductive coupling so instruments with excessive bandwidth for the task would not be useful. So I focused on a much simpler AD converter – a USB sound card.

The average sound card (USB or not) can take analog measurements 48000-96000 timer per second, with 16-bits of dynamic range(each 16 bit reading can range from 0 to 65535 or steps of 0.000015 which is quite remarkable really). That’s perfect if you want to measure frequencies below 24Khz or take voltage readings a few times a second (up to 60 times a second).

To put that into perspective, a 60Hz "signal" such as that coming out of the power outlet, can be measured with incredible accuracy - 800 samples per revolution (picture a sine wave made up of 800 dots). At a sample rate of 96kHz that number doubles and a single 60Hz waveform is now sampled 1600 times in just 1/60th (0.017) of a second. The 16 bit dynamic range is the big selling point here, because it allows us to measure extremely minute distortions to the waveform. These distortions are where all the interesting stuff happens. Not to mention, quality usb sound cards can be had for $5. Now we’re ready to track down anomalies. 

Anomalies and distortions are also referred to as harmonics. Harmonics come from many sources, mainly from things drawing current and causing slight variations to the waveform.

The fundamental 60Hz frequency itself is not distorted by local events however, that comes straight from the source(s). On a typical power grid there can be dozens of plants adding supply to the system. All of these generation plants have to add power at precisely the same frequency as those already on the grid, otherwise their generator short circuits due to phase mismatch.


examples/acpq.py
#!/usr/bin/env python


import argparse

# for config.yml
import yaml
# for placing HTTPs requests to EM7 and Scoreboard.
import requests
from requests.auth import HTTPBasicAuth
# For exit codes
import sys

import datetime

from gnuradio import audio
from gnuradio import filter
from gnuradio import blocks
from gnuradio import eng_notation
from gnuradio import gr
from gnuradio.eng_option import eng_option
from gnuradio.filter import firdes
from optparse import OptionParser

from gnuradio import analog
from gnuradio import blocks
from gnuradio import eng_notation
from gnuradio import gr
from gnuradio.eng_option import eng_option
from gnuradio.filter import firdes
from optparse import OptionParser

from elasticsearch import Elasticsearch

elasticSearch = {}
import json
import time
import socket
import certifi

import threading
import time

# Custom Blocks
import powerquality

SETTINGS = {}
ARGS = {}
# To be set to getEpochMillis() soon enough.
startTimeSeconds = 0


parser = argparse.ArgumentParser(prog='acpq',description="AC Power Quality Monitoring Script")
parser.add_argument('-c', '--config', help='Configuration File Path & Filename',required=False, default="./config.yml")
parser.add_argument('-p', '--port', help='Raw Samples Service - Tap Port',required=False, default="5555")
parser.add_argument('-D', '--debug', help='Enable debug messages to stdout',required=False)


def parseArgs():
    global ARGS

    # Parse commandline args. Prints help/usage and terminate application if anything goes wrong
    ARGS = parser.parse_args()
    print "Parsed commandline args. They are: " + str(ARGS)
    # Sanity Checks
    # If something doesn't check out, run argsFailed()


# Come here if parsing of commandline args failed, or if they are in some way invalid.
def argsFailed():
    parser.print_help()
    sys.exit()

# Returns a 2-digit month number
def getMonthString():
    dateMonthNumberString = "" + str(datetime.datetime.today().month)
    dateMonthNumberString = dateMonthNumberString.zfill(2)
    return dateMonthNumberString

def getYearString():
    dateYearNumberString = "" + str(datetime.datetime.today().year)
    return dateYearNumberString


def readSettings(configFileName):
    print "Loading settings from " + configFileName
    # Read settings from configuration file
    with open(configFileName, "r") as myfile:
        fileContents = myfile.read()
    print "Settings loaded."
    return yaml.load(fileContents)


def getHostName():
    return socket.gethostname()

def getEpochMillis():
    return long(time.time()*1000)

# Returns number of seconds since application started up.
def getRuntimeSeconds():
    global startTimeSeconds
    if (startTimeSeconds == 0):
        print "WARNING: initializing startTimeSeconds for the first time."
        startTimeSeconds = getEpochMillis() / 1000
    return (getEpochMillis()/1000) - startTimeSeconds

def getUniqueLogID():
    return "" + str(getEpochMillis())

def getProbeType():
    return SETTINGS["probe"]["probe_model"]

class testVolts(gr.top_block):
    def __init__(self):
        gr.top_block.__init__(self, "Testvolts")

        ##################################################
        # Variables - Some get overridden after block definition before connections.
        ##################################################
        self.samp_rate = samp_rate = 48000
        # Initial wavelength assumption, which must be calibrated carefully.
        self.fundamental_wavelength_samples = 800
        # self.probed_rms_volts = probed_rms_volts = 0

        # Things from config file
        # self.avg_length = avg_length = 10000
        self.probe_interval = probe_interval = float(SETTINGS["display"]["probe_interval"])
        self.rms_alpha = rms_alpha = 1

        # Calibration (Conversion) parameters, to convert from sample domain to volts/current domain.
        self.average_rms_slope =     float(SETTINGS["calibration"]["average_rms_slope"])
        self.average_rms_intercept = float(SETTINGS["calibration"]["average_rms_intercept"])

        # RMS Block parameters
        self.voltage_calculation_rms_alpha = float(SETTINGS["calibration"]["voltage_calculation_rms_alpha"])

        # Averaging block parameters
        self.voltage_rms_average_length =  int(SETTINGS["calibration"]["voltage_rms_average_length"])
        self.voltage_rms_average_scale =  float(SETTINGS["calibration"]["voltage_rms_average_scale"])
        self.voltage_rms_max_iter = int(SETTINGS["calibration"]["voltage_rms_max_iter"])

        # Input source parameters.
        self.audioSource = SETTINGS["input"]["input_type"]

        # Frequency calculation parameters
        self.freq_getfreq_alpha = float(SETTINGS["calibration"]["freq_getfreq_alpha"])
        self.freq_interpolation = float(SETTINGS["calibration"]["freq_interpolation"])
        self.freq_interpolation_multiplier = float(SETTINGS["calibration"]["freq_interpolation_multiplier"])

        # ELK send queue. Normally empty but if things are failing, data will queue until it can be sent.
        self.elk_send_queue = []

        self.logTemplate = {
            "logger": {
                "name":     SETTINGS["logging"]["logger_name"],
                "hostname": getHostName(),
                "model":     SETTINGS["logging"]["logger_model"],
                "location": SETTINGS["logging"]["logger_location"]
            },

            "probe": {
                "name":     SETTINGS["probe"]["probe_name"],
                "model":     SETTINGS["probe"]["probe_model"],
                "location": SETTINGS["probe"]["probe_location"],
            },

        }

        self.current_readings = self.getDefaultReadings()


        ##################################################
        # Define the blocks
        ##################################################
        self.probe_rms_volts = blocks.probe_signal_f()
        self.probe_avgwave = blocks.probe_signal_f()


        # Gather probe readings periodically.
        proball_thread = threading.Thread(target=self.probe_all_thread)
        proball_thread.daemon = True
        proball_thread.start()

        debug_print_thread = threading.Thread(target=self.periodic_logging_thread)
        debug_print_thread.daemon = True
        debug_print_thread.start()


        # Sound card and wav file variables will be instantiated shortly, during block connection.


        # AC Volts
        self.blocks_throttle_0 = blocks.throttle(gr.sizeof_float * 1, samp_rate, True)
        self.blocks_rms_volts = blocks.rms_ff(self.voltage_calculation_rms_alpha)
        self.blocks_moving_average_volts = blocks.moving_average_ff(self.voltage_rms_average_length, self.voltage_rms_average_scale, self.voltage_rms_max_iter)

        # AC Frequency
        self.upsampler = filter.fractional_resampler_ff(0, self.freq_interpolation)
        self.wavfile_input = blocks.wavfile_source('./samples/sample.wav', False)
        self.getfreq_block = powerquality.getfreqcpp(self.freq_getfreq_alpha)

        # Streaming support - opens a TCP port to which external applications can connect and get a tapped feed of raw unaltered samples.
        print "Initializing streaming server blocks."
        self.blocks_stream_to_tagged_stream = blocks.stream_to_tagged_stream(
            gr.sizeof_float,
            1,
            SETTINGS["networking_tap1"]["tagged_stream_packet_length"],
            SETTINGS["networking_tap1"]["length_tag_name"])
        self.blocks_tagged_stream_to_pdu = blocks.tagged_stream_to_pdu(
            blocks.float_t,
            SETTINGS["networking_tap1"]["length_tag_name"])
        self.blocks_socket_pdu = blocks.socket_pdu(
            "TCP_SERVER",
            SETTINGS["networking_tap1"]["socket_pdu_bind_host"],
            str(ARGS.port),
            SETTINGS["networking_tap1"]["socket_pdu_mtu"],
            SETTINGS["networking_tap1"]["socket_pdu_nodelay"])
        # Connect the streaming blocks
        # self.msg_connect((self.blocks_tagged_stream_to_pdu,"pdus"),(self.blocks_socket_pdu,"pdus"))
        self.msg_connect((self.blocks_tagged_stream_to_pdu,"pdus"),(self.blocks_socket_pdu,"pdus"))
        self.connect((self.blocks_stream_to_tagged_stream,0),(self.blocks_tagged_stream_to_pdu,0))


        self.blocks_multiply_const_vxx_0 = blocks.multiply_const_vff((-1, ))
        self.blocks_delay_0 = blocks.delay(gr.sizeof_float*1, self.fundamental_wavelength_samples)
        self.blocks_add_xx_0 = blocks.add_vff(1)



        ### Calibrate the blocks based on settings from config file
        self.set_rms_alpha(float(SETTINGS["calibration"]["voltage_calculation_rms_alpha"]))

        ##################################################
        # RMS Volts Connections
        ##################################################

        self.connect((self.blocks_moving_average_volts, 0), (self.probe_rms_volts, 0))
        self.connect((self.blocks_rms_volts, 0), (self.blocks_moving_average_volts, 0))

        # Set up a variable with a default value. Below we'll override the value depending on whether we are reading from file or audio device.
        self.sourceOfSamplesBlock = self.blocks_throttle_0

        # Select input source - wav file or sound card.
        if (self.audioSource == "wav"):
            # Variable source type: wav file
            self.input_wav_filename = SETTINGS["input"]["input_wav_filename"]
            self.wavfile_input = blocks.wavfile_source(self.input_wav_filename, True)

            # wav input requires throttle block. Then we connect things to the throttle block directly instead of the wav source.
            self.connect((self.wavfile_input, 0), (self.blocks_throttle_0, 0))

            self.sourceOfSamplesBlock = self.blocks_throttle_0

        else:
            # Variable source type: sound device.
            self.soundcard_device_name = ""
            if (SETTINGS["input"]["input_sound_device"]):
                self.soundcard_device_name = SETTINGS["input"]["input_sound_device"]

            self.soundcard_input = audio.source(samp_rate, self.soundcard_device_name, True)

            # no throttle block needed in the case of sound card

            self.sourceOfSamplesBlock = self.soundcard_input



        # Volts flow
        print "CONNECTIONS: Defining the voltage calculation flow"
        self.connect((self.sourceOfSamplesBlock , 0), (self.blocks_rms_volts, 0))
        # Hz flow
        print "CONNECTIONS: Defining the frequency calculation flow"
        self.connect((self.sourceOfSamplesBlock , 0), (self.upsampler, 0))
        # TCP port Tap flow
        print "CONNECTIONS: Connecting TCP/IP port tap flow"
        self.connect((self.sourceOfSamplesBlock , 0), (self.blocks_stream_to_tagged_stream, 0))


        # Streaming flow - add+accumulator -> tcp server.
        # TODO: This is processor intensive, so make it selectable from commandline option.
        # TODO: Actually, due to the lack of a TCP WAV streaming block, we should just stream baseband unaltered and let a separate process decode and also do data logging and things like that.
        if (1==0):
            print "CONNECTIONS: Creating a TCP server to enable access to the audio filter"
            self.connect((self.blocks_add_xx_0, 0), (self.blocks_tcp_server_sink_0, 0))
            print "CONNECTIONS: Setting up the audio filtering flow"
            self.connect((self.blocks_delay_0, 0), (self.blocks_multiply_const_vxx_0, 0))
            self.connect((self.blocks_multiply_const_vxx_0, 0), (self.blocks_add_xx_0, 1))
            self.connect((self.sourceOfSamplesBlock, 0), (self.blocks_add_xx_0, 0))
            self.connect((self.sourceOfSamplesBlock, 0), (self.blocks_delay_0, 0))



        ##################################################
        # AC Frequency connections
        ##################################################
        self.connect((self.getfreq_block, 0), (self.probe_avgwave, 0))
        self.connect((self.upsampler, 0), (self.getfreq_block, 0))

    def getDefaultReadings(self):

        ret = {
            # "timespanmillis": -1,
            "frequency": -1,
            "frequencymin": 60,
            "frequencymax": 60,
            "frequencychangecount":0,
        }

        if (getProbeType() == "RMSVOLTS"):
            ret["rmsvolts"] = -1
            ret["rmsvoltsmin"] = 999
            ret["rmsvoltsmax"] = -1
            ret["numvoltagespikes"] = -1
            ret["numvoltagesags"] = -1

        if (getProbeType() == "RMSCURRENT"):
            ret["rmscurrent"]   = -1
            ret["rmscurrentmin"]= -1
            ret["rmscurrentmax"] =-1



        return ret



    # TODO: Queueing. Keep it simple - create an object, key=getUniqueLogID, value=dumps (same as body).
    def sendCurrentReadingsToELK(self):
        fullMessage = self.logTemplate
        fullMessage["@timestamp"] = getEpochMillis()
        fullMessage["reading"] = self.current_readings

        # TODO: IF we start calling this function many times per second then it would make sense to not calculate the index every time.
        # 'power_quality-2017.07'
        #          power_quality-    2017               .    07

        # # Sanity checks
        # if self.current_readings["frequency"] < 10:
        #     print "WARNING: sendCurrentReadingsToELK() is rejecting one message due to absurd frequency reading of " + str(self.current_readings["frequency"])
        #     return
        # update: lifted the above sanity chceck as the getfreqcpp block has improved significantly in reliability.


        # Sanity check: Don't send results until after a brief warm-up period. averages, flow graph stuff, needs to settle in.
        if (getRuntimeSeconds() < 20):
            remainingSeconds = 20 - getRuntimeSeconds()
            print "WARMUP period not yet elapsed. Skipping one transmit interval to Elasticsearch. remaining seconds = " + str(remainingSeconds)
            return

        if (ARGS.debug):
            print "Sending to ES: " + json.dumps(fullMessage)

        self.elk_send_queue.append(fullMessage)

        if (len(self.elk_send_queue) > 1) or (ARGS.debug):
            print ("Queued one reading. Queue now contains " + str(len(self.elk_send_queue)) + " items.")




    # "Run the queue" - send items to ELK.
    def sendQueuedReadingsToELK(self):

        if ("max_sends_per_cycle" in SETTINGS["logging"]):
            max_send_per_run = int(SETTINGS["logging"]["max_sends_per_cycle"])
        else:
            print ("WARNING: Config file is missing parameter logging.max_sends_per_cycle. Defaulting to 10.")
            max_send_per_run = 10

        indexName = "power_quality-" + getYearString() + "." + getMonthString()

        # Figure out how many items we'll be processing. no more than the allowed max.
        num_items_this_run = max_send_per_run
        if (len(self.elk_send_queue) < max_send_per_run):
            num_items_this_run = len(self.elk_send_queue)


        for i in range (num_items_this_run):
            # read the next item in the queue
            current_item = self.elk_send_queue[0]

            # Try sending it
            try:
                # Default: elk=enabled, but can be set to False in the config file to disable sending of data to elk.
                elk_enabled = True
                if ("elk_enabled" in SETTINGS["logging"]):
                    elk_enabled = SETTINGS["logging"]["elk_enabled"]

                if (elk_enabled == True):
                    elasticSearch.index(index=indexName, doc_type='post', id=getUniqueLogID(), body=json.dumps(current_item ))
                else:
                    print ("elk_enabled=False so not sending this to ELK: " + json.dumps(current_item))

                # TODO Check within the response to look for some sign that it was definitely certainly for sure successful.
                # successful send? Woohoo! pop it off the left side of the list.
                self.elk_send_queue.pop()
                if (len(self.elk_send_queue) > 1):
                    print ("Successfully sent one queued reading, there are " + str(len(self.elk_send_queue)) + " more left in the queue!")
            except:
                print("Error sending data to index. Unable to send queued readings at this time. elkError=", sys.exc_info()[0])
                # send was NOT successful. Bail out of this function so when connection is restored we start at the beginning of the queue.
                return


    def get_samp_rate(self):
        return self.samp_rate

    def set_samp_rate(self, samp_rate):
        self.samp_rate = samp_rate
        self.blocks_throttle_0.set_sample_rate(self.samp_rate)

    def set_rms_alpha(self, newAlpha):
        print "Setting rms_volts alpha to " + str(newAlpha)
        self.rms_alpha = newAlpha
        self.blocks_rms_volts.set_alpha(self.rms_alpha)

    def get_probed_rms_volts(self):
        return self.probed_rms_volts

    # Apply the conversion factors specified in the config file to convert voltage sample values into actual voltage reading
    def convertSampleToReading(self, sampleReading):
        return sampleReading * self.average_rms_slope + self.average_rms_intercept

    def set_getAvgwaveLevel(self, getAvgwaveLevel):
        self.getAvgwaveLevel = getAvgwaveLevel


    def set_current_reading(self, reading, rawReading):
         # Keep track of min/max values observed. These get reset by reset_accumulators() each time after sending the readings to elasticsearch.

        if (getProbeType() == "RMSVOLTS"):

            if (reading < self.current_readings["rmsvoltsmin"]):
                self.current_readings["rmsvoltsmin"] = reading
            if (reading > self.current_readings["rmsvoltsmax"]):
                self.current_readings["rmsvoltsmax"] = reading

            # Set current actual reading values.
            self.current_readings["rmsvolts"] = reading
            self.current_readings["rmsvoltsprobereading"] = rawReading

        if (getProbeType() == "RMSCURRENT"):
            if (reading < self.current_readings["rmscurrentmin"]):
                self.current_readings["rmscurrentmin"] = reading
            if (reading > self.current_readings["rmscurrentmax"]):
                self.current_readings["rmscurrentmax"] = reading

            # Set current actual reading values.
            self.current_readings["rmscurrent"] = reading
            self.current_readings["rmscurrentprobereading"] = rawReading

    def set_current_frequency(self,hz):
        # Keep track of min/max values observed. These get reset by reset_accumulators() each time after sending the readings to elasticsearch.
        if (hz < self.current_readings["frequencymin"]):
            self.current_readings["frequencymin"] = hz
        if (hz > self.current_readings["frequencymax"]):
            self.current_readings["frequencymax"] = hz

        # Did the frequency change? TODO: better things with frequency changes.
        if (hz != self.current_readings["frequency"]):
            self.current_readings["frequencychangecount"] = int(self.current_readings["frequencychangecount"]) + 1

        # Set current actual reading values.
        self.current_readings["frequency"] = hz


    def periodic_logging_thread(self):
        while True:
            # print str(self.current_readings)
            self.sendCurrentReadingsToELK()
            self.reset_accumulators()
            # run the queue - at least one item (the current one) will be queued. If connection problems then more will be queued.
            self.sendQueuedReadingsToELK()
            # TODO: The following sleep should look at config file parameter for sleep duration.
            time.sleep(1)

    def probe_all_thread(self):
        while True:
            self.probe_frequency()
            self.probe_rms_reading()
            time.sleep(1.0 / (self.probe_interval))

    def probe_frequency(self):
        frequency = self.probe_avgwave.level()
        try:
            self.set_getAvgwaveLevel(frequency)
            if frequency > 0:
                myFreq = (self.samp_rate * self.freq_interpolation_multiplier) / frequency
                self.set_current_frequency(myFreq)
        except AttributeError:
            pass

    def probe_rms_reading(self):
        rmsreading = self.probe_rms_volts.level()
        self.set_current_reading(self.convertSampleToReading(rmsreading), rmsreading)

    # Run this each time the readings are sent to elasticsearch, to reset any accumulator variables.
    def reset_accumulators(self):
        self.current_readings["frequencychangecount"] = 0
        self.current_readings["frequencymax"] = self.current_readings["frequency"]
        self.current_readings["frequencymin"] = self.current_readings["frequency"]

        if (getProbeType() == "RMSVOLTS"):
            self.current_readings["rmsvoltsmin"] = self.current_readings["rmsvolts"]
            self.current_readings["rmsvoltsmax"] = self.current_readings["rmsvolts"]

        if (getProbeType() == "RMSCURRENT"):
            self.current_readings["rmscurrentmin"] = self.current_readings["rmscurrent"]
            self.current_readings["rmscurrentmax"] = self.current_readings["rmscurrent"]


# This gets called after config file and commandline args have been parsed. Do things like initializing variables.
def preStartTasks():
    global elasticSearch

    startTimeSeconds = getEpochMillis() / 1000

    # Initialize Elasticsearch Logging.
    esTarget = SETTINGS["logging"]["logtarget"]
    print "Initializing Elasticsearch. Endpoint=" + esTarget
    elasticSearch = Elasticsearch(esTarget,ca_certs=certifi.where())



def main(top_block_cls=testVolts, options=None):
    global SETTINGS

    # Parse commandline arguments
    print "Note! This application shares the config.yml file with the streamer."
    parseArgs()

    # Parse the config file
    SETTINGS = readSettings(ARGS.config)

    # Perform initial configuration tasks now that config data is available, before main processing starts up.
    preStartTasks()

    tb = top_block_cls()
    tb.start()
    tb.wait()


if __name__ == '__main__':
    main()
